### PR TITLE
Fix DecomposeMultiDimSqueeze callback failure

### PR DESCRIPTION
**Pull request by meenakshiramanathan1**
Fixes [#642](https://github.com/tenstorrent/tt-forge-fe/issues/642)

### CODE REVIEW

File: python/tvm/relay/op/contrib/forge/forge_passes.py [hunk 1]
```diff
@@ -1883,6 +1883,8 @@ def __init__(self):
     def callback(self, pre, post, node_map):
         act = node_map[self.act][0]
         axis = post.attrs.axis
+        if any([isinstance(dim, tvm.tir.expr.Any) for dim in pre.checked_type.shape]):
```
Comment (dgolubovicTT) [inline python/tvm/relay/op/contrib/forge/forge_passes.py:1886]: Can you give some context in PR why the change?

Comment (meenakshiramanathan1) [reply python/tvm/relay/op/contrib/forge/forge_passes.py:1886]: @dgolubovicTT , this change fixes [#642](https://github.com/tenstorrent/tt-forge-fe/issues/642), the current status of this model is mentioned in the description [here](https://github.com/tenstorrent/tt-forge-fe/pull/865).

Comment (meenakshiramanathan1) [reply python/tvm/relay/op/contrib/forge/forge_passes.py:1886]: Yes, in that case we skip removal of squeeze containing dynamic shapes, warning can be added. Squeeze op is supported in mlir as mentioned [here](https://github.com/tenstorrent/tt-forge-fe/blob/8ed75136aba734c23300cd92a73e71192dbbcf46/forge/csrc/passes/lower_to_mlir.cpp#L587). 
Densenet model is failing with unsupported ops - argwhere and scatter_elements after this change

Comment (dgolubovicTT) [reply python/tvm/relay/op/contrib/forge/forge_passes.py:1886]: Ok, can you add a single squeeze test that will test this case, just to be sure everything after tvm will work.

Comment (meenakshiramanathan1) [reply python/tvm/relay/op/contrib/forge/forge_passes.py:1886]: [These](https://github.com/tenstorrent/tt-forge-fe/blob/79e186f17dd5aa0e14a3a358a66de47f503c13ed/forge/test/mlir/test_ops.py#L981) test cases for squeeze op were added as a part of this [PR](https://github.com/tenstorrent/tt-forge-fe/pull/528/files#diff-9094f8f34c15528e4fad062984363b21b8e2fdd4b19c6d49878c16cebea33484) when DecomposeMultiDimSqueeze callback was added.

Comment (dgolubovicTT) [reply python/tvm/relay/op/contrib/forge/forge_passes.py:1886]: But is any of them actually testing this code path where this condition is not satisfied:
```
 if any([isinstance(dim, tvm.tir.expr.Any) for dim in pre.checked_type.shape]):
```

Comment (meenakshiramanathan1) [reply python/tvm/relay/op/contrib/forge/forge_passes.py:1886]: I will try this and let you know.

**Comment by nvukobratTT** [comment]
@meenakshiramanathan1 can you provide a bit more context to this PR. Here are few questions I'm interested in:
1. The one @dgolubovicTT already asked, are we testing this path? If no, let's try to reproduce this with simple op test.
2. What are the results with and without your fix? 
    - Before the change, you end up with failure in `DecomposeMultiDimSqueeze `
    - What about when you add this fix? Howe this multi-dim squeeze is further lowered in FFE, and latter on into MLIR? Can you share some reportify graphs for few compile stages (both TVM and FFE), as well a TTIR? 
    - If you're unable to create simple unit test to cover this one, let's see screenshots from the original model.

**Comment by meenakshiramanathan1** [comment]
> @meenakshiramanathan1 can you provide a bit more context to this PR. Here are few questions I'm interested in:
> 
> 1. The one @dgolubovicTT already asked, are we testing this path? If no, let's try to reproduce this with simple op test.
> 2. What are the results with and without your fix?
>    
>    * Before the change, you end up with failure in `DecomposeMultiDimSqueeze `
>    * What about when you add this fix? Howe this multi-dim squeeze is further lowered in FFE, and latter on into MLIR? Can you share some reportify graphs for few compile stages (both TVM and FFE), as well a TTIR?
>    * If you're unable to create simple unit test to cover this one, let's see screenshots from the original model.

1. Tried testing this path with dynamic inputs but in the all cases of trying to create dynamic inputs , it is ending up with formation of argwhere which is not supported in forge yet
For example,
```
def test_dynamic_multi_squeeze(test_device):
    class Multi_Squeeze(torch.nn.Module):
        def __init__(self):
            super().__init__()

        def forward(self, input):
            mask = input > 5
            indices = torch.nonzero(mask)
            squeezed_tensor = indices.squeeze(0)
            return squeezed_tensor
```
Here non_zero is mapped to argwhere which was one of the cases which was tried, so if we want to test this out further , argwhere op tvm decomposition should be brought in.

2. With my fix , model is ending up with argwhere unsupported and without my fix, it would end up with [this](https://github.com/tenstorrent/tt-forge-fe/issues/642) error.

3. Screenshots from original model which is causing dynamic shapes:
Screenshot 2024-12-28 at 9 01 58 AM

**Comment by nvukobratTT** [comment]
@meenakshiramanathan1 

> Tried testing this path with dynamic inputs but in the all cases of trying to create dynamic inputs , it is ending up with formation of argwhere which is not supported in forge yet

Thanks for the details! We're pushing the effort on outlining what and how our indexing should work. 

> Here non_zero is mapped to argwhere which was one of the cases which was tried, so if we want to test this out further , argwhere op tvm decomposition should be brought in.
> 
> Before fix , graps are dumped till relay passes.

Can you send actually screenshot of the graph? I'm interested in graph structure changes, rather than compile depth.

**Comment by meenakshiramanathan1** [comment]
> @meenakshiramanathan1
> 
> > Tried testing this path with dynamic inputs but in the all cases of trying to create dynamic inputs , it is ending up with formation of argwhere which is not supported in forge yet
> 
> Thanks for the details! We're pushing the effort on outlining what and how our indexing should work.
> 
> > Here non_zero is mapped to argwhere which was one of the cases which was tried, so if we want to test this out further , argwhere op tvm decomposition should be brought in.
> > Before fix , graps are dumped till relay passes.
> 
> Can you send actually screenshot of the graph? I'm interested in graph structure changes, rather than compile depth.

Screenshot 2025-01-09 at 3 41 16 PM
Screenshot 2025-01-09 at 3 43 59 PM

**Review by nvukobratTT — APPROVED** [review]
